server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remote/delivery/DeliveryRunnable.java

```diff
@@ -52,77 +55,65 @@
 
-    @Override
-    public void run() {
+    public void start() {
+        disposable = queue.deQueue()
+            .publishOn(Schedulers.newParallel("RemoteDelivery", configuration.getWorkersThreadCount()))
+            .flatMap(this::runStep)
+            .onErrorContinue(((throwable, nothing) -> LOGGER.error("Exception caught in RemoteDelivery", throwable)))
+            .subscribeOn(Schedulers.elastic())
+            .subscribe();
+    }
+
+    private Mono<Void> runStep(MailQueue.MailQueueItem queueItem) {
+        TimeMetric timeMetric = metricFactory.timer(REMOTE_DELIVERY_TRIAL);
         try {
-            while (!Thread.interrupted() && !isDestroyed.get()) {
-                runStep();
-            }
+            return processMail(queueItem);
+        } catch (Throwable e) {
+            return Mono.error(e);
         } finally {
-            // Restore the thread state to non-interrupted.
-            Thread.interrupted();
+            timeMetric.stopAndPublish();
         }
     }
 
-    private void runStep() {
-        TimeMetric timeMetric = null;
-        try {
-            // Get the 'mail' object that is ready for deliverying. If no message is
-            // ready, the 'accept' will block until message is ready.
-            // The amount of time to block is determined by the 'getWaitTime' method of the MultipleDelayFilter.
-            MailQueue.MailQueueItem queueItem = queue.deQueue();
-            timeMetric = metricFactory.timer(REMOTE_DELIVERY_TRIAL);
-            Mail mail = queueItem.getMail();
-
-            try {
-                if (configuration.isDebug()) {
-                    LOGGER.debug("{} will process mail {}", Thread.currentThread().getName(), mail.getName());
-                }
-                attemptDelivery(mail);
-                LifecycleUtil.dispose(mail);
-                mail = null;
-                queueItem.done(true);
-            } catch (Exception e) {
-                // Prevent unexpected exceptions from causing looping by removing message from outgoing.
-                // DO NOT CHANGE THIS to catch Error!
-                // For example, if there were an OutOfMemory condition caused because
-                // something else in the server was abusing memory, we would not want to start purging the retrying spool!
-                LOGGER.error("Exception caught in RemoteDelivery.run()", e);
-                LifecycleUtil.dispose(mail);
-                queueItem.done(false);
-            }
+    private Mono<Void> processMail(MailQueue.MailQueueItem queueItem) throws MailQueue.MailQueueException {
+        Mail mail = queueItem.getMail();
 
-        } catch (Throwable e) {
-            if (!isDestroyed.get()) {
-                LOGGER.error("Exception caught in RemoteDelivery.run()", e);
-            }
+        try {
+            LOGGER.debug("will process mail {}", mail.getName());
+            attemptDelivery(mail);
+            queueItem.done(true);
+            return Mono.empty();
```

Comment:
Again, the method signature let me understand this is actually a reactor handled execution, but it turns out we do execute everything synchronously then return a fake result.

This is conter intuitive. Why don't we just wrap all of this synchronous process in a `Mono.fromRunnable` ?